test/TestProjects/SerializationCustomization/Generated/Models/AlwaysInitializeTestModel.cs

```diff
@@ -0,0 +1,65 @@
+// Licensed under the MIT License.
+
+// <auto-generated/>
+
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SerializationCustomization.Models
+{
+    /// <summary> The AlwaysInitializeTestModel. </summary>
+    public partial class AlwaysInitializeTestModel
+    {
+        /// <summary> Initializes a new instance of AlwaysInitializeTestModel. </summary>
+        /// <param name="requiredList"> . </param>
+        /// <param name="requiredAlwaysInitializeList"> . </param>
+        /// <param name="requiredAlwaysInitializeObject"> . </param>
+        public AlwaysInitializeTestModel(IEnumerable<Item> requiredList, IEnumerable<Item> requiredAlwaysInitializeList, Item requiredAlwaysInitializeObject)
+        {
+            if (requiredList == null)
+            {
+                throw new ArgumentNullException(nameof(requiredList));
+            }
+            if (requiredAlwaysInitializeList == null)
+            {
+                throw new ArgumentNullException(nameof(requiredAlwaysInitializeList));
+            }
+            if (requiredAlwaysInitializeObject == null)
+            {
+                throw new ArgumentNullException(nameof(requiredAlwaysInitializeObject));
+            }
+
+            AlwaysInitializeList = new List<Item>();
+            RequiredList = requiredList.ToArray();
+            RequiredAlwaysInitializeList = requiredAlwaysInitializeList.ToArray();
+            RequiredAlwaysInitializeObject = requiredAlwaysInitializeObject;
+            AlwaysInitializeObject = new Item();
+        }
+
+        /// <summary> Initializes a new instance of AlwaysInitializeTestModel. </summary>
+        /// <param name="alwaysInitializeList"> . </param>
+        /// <param name="requiredList"> . </param>
+        /// <param name="requiredAlwaysInitializeList"> . </param>
+        /// <param name="defaultList"> . </param>
+        /// <param name="requiredAlwaysInitializeObject"> . </param>
+        /// <param name="alwaysInitializeObject"> . </param>
+        /// <param name="defaultObject"> . </param>
+        internal AlwaysInitializeTestModel(IList<Item> alwaysInitializeList, IList<Item> requiredList, IList<Item> requiredAlwaysInitializeList, IList<Item> defaultList, Item requiredAlwaysInitializeObject, Item alwaysInitializeObject, Item defaultObject)
+        {
+            AlwaysInitializeList = alwaysInitializeList ?? new List<Item>();
+            RequiredList = requiredList;
+            RequiredAlwaysInitializeList = requiredAlwaysInitializeList ?? new List<Item>();
+            DefaultList = defaultList;
+            RequiredAlwaysInitializeObject = requiredAlwaysInitializeObject ?? new Item();
+            AlwaysInitializeObject = alwaysInitializeObject ?? new Item();
+            DefaultObject = defaultObject;
+        }
+        public IList<Item> RequiredList { get; }
+        public IList<Item> DefaultList { get; set; }
```

Comment:
This still violates .NET Guidelines. Collection properties should never be settable.

Reply:
Guidelines don't cover all cases that generator are used for so they would need to change to accomodate.

Reply:
The guidelines are there for consistent behavior. If collection properties can sometimes be null, that's even worse. As I understand, making them settable originated because of the null-or-empty problem. I think we should focus on solving that (even if it requires "hints" like this PR) and have a consistent public API. @KrzysztofCwalina, thoughts? If we really want to entertain "sometimes-null" collection properties or collection properties that can be set - completely breaking any stored (even temporarily in a call path) references to them - this needs to be a larger discussion; though, perhaps not blocking this PR. At least this is a great start to making it possible to opt into the guidelines.

Reply:
What I don't know how to solve yet is there are 200 management plane libraries and no resources to add hints vs 2 data plane with plenty resources. This is the reason I'm reluctant to flip the switch right away and want to gather more data.

I agree we need a larger discussion, hope API review of storage management might be that.